Add support for provisioning macOS jenkins agents

nuclearsandwich opened the pull request:
This PR adds a new recipe for creating macOS agents.

Right now it's actually quite basic because the low-level macOS host setup is hostile to automation and has drifted from previous methods to set things up via `defaults write`. The machines that this was developed on and shipped for were provisioned by hand using a privately documented process.

I do not think that the low level setup is required for this PR to have value so while I hope that we'll find ways to maintainably audit the initial machine setup that lack does not block this PR from review and merge.

nuclearsandwich committed:
Start macos recipe.

nuclearsandwich committed:
Try to resolve error on deploy.

nuclearsandwich committed:
Give up on managing users and try setting a default.

nuclearsandwich committed:
This default did not apply. See if this helps.

nuclearsandwich committed:
Try running the command directly.

nuclearsandwich committed:
Swtich gears to Jenkins setup.

Base level macOS setup is too trial-and-error and will have to be done
by hand for now.

nuclearsandwich committed:
Only download and install the package if it is not already installed.

nuclearsandwich committed:
Add launchd service for running Jenkins.

nuclearsandwich committed:
Drop group ownership.

macOS does not create a group for each user.

nuclearsandwich committed:
Use launchd resource to create and start jenkins-agent.

nuclearsandwich committed:
Copy labels array so it can be mutated.

nuclearsandwich committed:
Use log directory within jenkins user directory.

nuclearsandwich committed:
Create jenkins-agent directory and move workspace root there.

nuclearsandwich committed:
Specify correct group for swarm client ownership.

nuclearsandwich committed:
Enhance description and add architecture label.

Anticipating future Apple Silicon macs.

nuclearsandwich committed:
Put argument flag and value on same line.

nuclearsandwich committed:
Add environment variables to agent.

nuclearsandwich committed:
Install homebrew cask and run it at Jenkins login.

nuclearsandwich committed:
Install xquartz via direct download rather than homebrew cask.

We want homebrew operations to run via the Jenkins user to keep
permissions in good order but the Chef executor cannot do so without a
password prompt which fails. So this is just easier.

The biggest downside is that it does not monitor/maintain updates but
neither does homebrew-cask by default either so :shrug:.

nuclearsandwich committed:
One -e is required per variable.

And --env nor -env seem to work.

nuclearsandwich committed:
Fix XQuartz program path.

nuclearsandwich commented on recipes/macos.rb:
```diff
@@ -0,0 +1,114 @@
+# Install xquartz
+remote_file "/tmp/xquartz.pkg" do
+  source "https://github.com/XQuartz/XQuartz/releases/download/XQuartz-2.8.5/XQuartz-2.8.5.pkg"
+  not_if "pkgutil --pkg-info org.xquartz.X11"
+end
```
@scpeters I ended up installing XQuartz via pkg download rather than homebrew cask because the homebrew cask approach does not work correctly when the user chef runs as isn't the `jenkins` user we want to be the homebrew owner.

Since this was a cask rather than a formula the main trade-off is that we have to manually update the XQuartz version in this recipe when we want to change the installed version (and the recipe only checks that the application is installed not that the version matches our desired).
However, since homebrew also does not support updating casks that have auto-update mechanisms I don't think this is a deal breaker.

scpeters replied:
sounds good to me

nuclearsandwich commented on recipes/macos.rb:
```diff
@@ -0,0 +1,114 @@
+  recursive true
+end
+
+launchd "org.xquartz.X11.plist" do
+  path "/Users/jenkins/Library/LaunchAgents/org.xquartz.X11.plist"
+  keep_alive true
+  run_at_load true
+  working_directory "/Users/jenkins"
+  process_type "Interactive"
+  program "/Applications/Utilities/XQuartz.app/Contents/MacOS/X11"
+  action [:create, :enable]
+end
```
Also a deviation from our previous setup instructions, I'm running xquartz as a launch agent rather than as a login item. The advantage of this is that xquartz will be restarted on crash. Which I verified a few times by killing the x11 process and watching it spawn a new one. Should cut down on human interventions required 🤞🏼

scpeters replied:
👍

nuclearsandwich commented on recipes/macos.rb:
```diff
@@ -0,0 +1,114 @@
+end
+
+
+# Install java
+remote_file "/tmp/jdk8.pkg" do
+  source "https://github.com/adoptium/temurin8-binaries/releases/download/jdk8u362-b09/OpenJDK8U-jdk_x64_mac_hotspot_8u362b09.pkg"
+  not_if "pkgutil --pkg-info net.temurin.8.jdk"
+end
+
+execute "install java" do
+  command "installer -pkg /tmp/jdk8.pkg -target /"
+  not_if "pkgutil --pkg-info net.temurin.8.jdk"
+end
```
I've opted not to prematurely add logic to support JDK8 vs JDK11, which is an upcoming transition required on the build farm. I have chosen to keep this PR straightforward and add the JDK11 support in a follow-up.

scpeters commented:
I see that `brew doctor` is complaining about the homebrew sbin folder not being in the `PATH`:

* [![Build Status](https://build.osrfoundation.org/buildStatus/icon?job=ignition_common4-install_bottle-homebrew-amd64&build=713)](https://build.osrfoundation.org/job/ignition_common4-install_bottle-homebrew-amd64/713/) https://build.osrfoundation.org/job/ignition_common4-install_bottle-homebrew-amd64/713/


```
+ brew doctor
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: Homebrew's "sbin" was not found in your PATH but you have installed
formulae that put executables in /usr/local/sbin.
Consider setting your PATH for example like so:
  echo 'export PATH="/usr/local/sbin:$PATH"' >> ~/.zshrc
+ echo MARK_AS_UNSTABLE
```

nuclearsandwich committed:
Document manual steps which are not yet automated.

There was friction in automating these steps with chef which we
haven't worked through.

claraberendsen commented on recipes/macos.rb:
```diff
@@ -1,3 +1,44 @@
+# Verify wifi and bluetooth are disabled, which should already be true for
+# our hosted machines.
+
+# Disable Spotlight indexing. It's worth doing globally but at the very least
```
[This](https://github.com/Microsoft/macos-cookbook/blob/master/documentation/resource_spotlight.md) microsoft chef resource could be useful for the automation of this.

nuclearsandwich replied:
Agreed! I am strongly committed to avoiding cookbook dependencies but this one may be worth adopting in order to deal more directly with macos intricacies that are hard to model within our own recipes. However, I think that merging this as-is and automating more of the set up in further PRs is the way to go.

j-rivero commented on recipes/macos.rb:
```diff
@@ -0,0 +1,155 @@
+end
+
+directory "/Users/jenkins/Library/LaunchAgents" do
+  owner "jenkins"
```
Nitpick: we could abstract the user and group in the attributes file and use the variable in this file replacing the name in multiple places. Feel free to skip at this point, not sure if it worth it.

j-rivero commented on recipes/macos.rb:
```diff
@@ -0,0 +1,155 @@
+    -url #{node['osrfbuild']['agent']['jenkins_url']}
+    -name #{agent_name}
+    -username #{jenkins_agent_user['username']}
+    -password #{jenkins_agent_user['password']}
```
Not familiar with Mac, but using the `password` argument directly here could imply a security problem if the process if visible for all users in the machine. I remember to have worked on this after a security review point that out in our infra. The [linux recipe uses `--passwordFile` ](https://github.com/osrf/osrf_jenkins_agent/blob/latest/templates/jenkins-agent.service.erb#L9 ) modified in https://github.com/osrf/osrf_jenkins_agent/pull/17

j-rivero commented on recipes/macos.rb:
```diff
@@ -0,0 +1,155 @@
+    -deleteExistingClients
+    -labels #{labels.join(' ')}
+    -e HOMEWBREW_FORCE_VENDOR_RUBY=1
+    -e MAKE_JOBS=8
```
Nitpick: we don't have a better way for defining this but maybe moving the same hardcoded value to an attribute can make people life easier if they need to modify it.

j-rivero commented on recipes/macos.rb:
```diff
@@ -0,0 +1,155 @@
+    -password #{jenkins_agent_user['password']}
+    -description #{description}
+    -mode exclusive
+    -executors 1
```
Nitpick: also a good candidate to be included in an attribute file.

j-rivero reviewed (commented):
Sorry for not reviewing it before. Mostly good to go, one suggestion I think that it is important about security and password.

Crola1702 commented on recipes/macos.rb:
```diff
@@ -0,0 +1,155 @@
+
+# Verify SSH and VNC remote access are enabled, which should already true for
+# our hosted machines.
+# Verify remote management is enabled _only_ for administrator
```
```suggestion
```

This can lead to VNC connection problems

claraberendsen committed:
add sonoma and sequoia support

Crola1702 committed:
[24.04]  MacOS: Use temurin21 instead of temurin8 (#49)

* Use temurin21 instead of temurin8

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

---------
Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>
Signed-off-by: Jose Luis Rivero <jrivero@honurobotics.com>

j-rivero committed:
Merge remote-tracking branch 'origin' into nuclearsandwich/osrfbuild-mac-bringup

Crola1702 committed:
Support arm mac (#54)

* Use curl to install Xquartz, install different temurin archs

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

* Use curl to install temurin also

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

* Fix syntax error

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

* Use curl for swarm client

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

* Don't download swarm client if it already exists

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

* Add comment on why using curl

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

* Add arch+os as label

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

* Add conditional to match the labels from homebrew bottles

See https://github.com/gazebo-tooling/release-tools/pull/1364#issuecomment-3202620472

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

---------

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

j-rivero commented:
Going to merge this PR as it is since it was used to deploy the new ARM agents by @Crola1702 . Want to follow up the problem reported in the review with the password being passed as a clear argument to the process list https://github.com/osrf/osrf_jenkins_agent/pull/38/files#r1202628021. I'll create a new PR for it.
